radix-engine-stores/src/hash_tree/tree_store.rs

```diff
@@ -0,0 +1,163 @@
+use std::collections::HashMap;
+
+use radix_engine_interface::api::types::SubstateId;
+use radix_engine_interface::crypto::Hash;
+
+/// A nibble (4 LSBs of the contained u8).
+#[derive(Clone, PartialEq, Eq, Hash)]
+pub struct Nib(pub u8);
+
+/// A sequence of nibbles (potentially not byte-aligned).
+#[derive(Clone, PartialEq, Eq, Hash)]
+pub struct Nibs(pub Vec<Nib>);
+
+impl Nibs {
+    /// Creates a new sequence from the given parts.
+    pub fn concat(&self, other: &Nibs) -> Nibs {
+        Nibs([self.0.as_slice(), other.0.as_slice()].concat())
+    }
+
+    /// Consumes the `nib_count` first nibbles of this sequence.
+    pub fn skip(mut self, nib_count: usize) -> Nibs {
+        self.0.drain(0..nib_count);
+        self
+    }
+}
+
+/// A unique key of a "physical" tree node, to be used in the storage.
+#[derive(Clone, PartialEq, Eq, Hash, Debug)]
+pub struct TreeNodeKey {
+    /// State version - starting with 1, incremented strictly by 1.
+    pub version: u64,
+    /// Nibbles, potentially not byte-aligned.
+    pub nib_prefix: Nibs,
+}
+
+/// A physical tree node, to be used in the storage.
+#[derive(Clone, PartialEq, Eq, Hash, Debug)]
+pub enum TreeNode {
+    /// Internal node - always metadata-only, as per JMT design.
+    Internal(TreeInternalNode),
+    /// Leaf node.
+    Leaf(TreeLeafNode),
+    /// An "empty tree" indicator, which may only be used as a root.
+    Null,
+}
+
+/// Internal node.
+#[derive(Clone, PartialEq, Eq, Hash, Debug)]
+pub struct TreeInternalNode {
+    /// Metadata of each existing child.
+    pub children: Vec<TreeChildEntry>,
+}
+
+/// Child node metadata.
+#[derive(Clone, PartialEq, Eq, Hash, Debug)]
+pub struct TreeChildEntry {
+    /// First of the remaining nibbles in the key.
+    pub nib: Nib,
+    /// State version at which this child's node was created.
+    pub version: u64,
+    /// Cached child hash (i.e. needed only for performance).
+    pub hash: Hash,
+    /// Cached child type indicator (i.e. needed only for performance).
+    pub is_leaf: bool,
+}
+
+/// Physical leaf node (which may represent a ReNode or a Substate).
+#[derive(Clone, PartialEq, Eq, Hash, Debug)]
+pub struct TreeLeafNode {
+    /// All the remaining nibbles in the _hashed_ `substate_id`.
+    pub nib_suffix: Nibs,
```

Comment:
But what if
```
A
|
X
```

Becomes:
```
A
|
B
| \
X Y
```
I don't think `X` gets rewritten by the tree - but now its suffix is incorrect.

Reply:
> So it must be written to the DB, because it will be expected by the parent under new db key.

Ohhh yes! That makes sense! Really appreciate you explaining it to me 👍 

I'm happy with either approach.